consensus/state_processing/src/per_epoch_processing/altair/participation_cache.rs

```diff
@@ -0,0 +1,409 @@
+    previous_epoch_participation: SingleEpochParticipationCache,
+    /// Caches the result of the `get_eligible_validator_indices` function.
+    eligible_indices: Vec<usize>,
+}
+
+impl ParticipationCache {
+    /// Instantiate `Self`, returning a fully initialized cache.
+    ///
+    /// ## Errors
+    ///
+    /// - The provided `state` **must** be an Altair state. An error will be returned otherwise.
+    pub fn new<T: EthSpec>(
+        state: &BeaconState<T>,
+        spec: &ChainSpec,
+    ) -> Result<Self, BeaconStateError> {
+        let current_epoch = state.current_epoch();
+        let previous_epoch = state.previous_epoch();
+
+        let num_previous_epoch_active_vals = state
+            .get_cached_active_validator_indices(RelativeEpoch::Previous)?
+            .len();
+        let num_current_epoch_active_vals = state
+            .get_cached_active_validator_indices(RelativeEpoch::Current)?
+            .len();
+
+        // Both the current/previous epoch participations are set to a capacity that is slightly
+        // larger than required. The difference will be due slashed-but-active validators.
+        let mut current_epoch_participation =
+            SingleEpochParticipationCache::new(num_current_epoch_active_vals, spec);
+        let mut previous_epoch_participation =
+            SingleEpochParticipationCache::new(num_previous_epoch_active_vals, spec);
+        // Contains the set of validators which are either:
+        //
+        // - Active in the previous epoch.
+        // - Slashed, but not yet withdrawable.
+        //
+        // Using the full length of `state.validators` is almost always overkill, but it ensures no
+        // reallocations.
+        let mut eligible_indices = Vec::with_capacity(state.validators().len());
+
+        // Iterate through all validators, updating:
+        //
+        // 1. Validator participation for current and previous epochs.
+        // 2. The "eligible indices".
+        //
+        // Care is taken to ensure that the ordering of `eligible_indices` is the same as the
+        // `get_eligible_validator_indices` function in the spec.
+        for (val_index, val) in state.validators().iter().enumerate() {
+            if val.is_active_at(current_epoch) {
+                current_epoch_participation.process_active_validator(
+                    val_index,
+                    state,
+                    RelativeEpoch::Current,
+                )?;
+            }
+
+            if val.is_active_at(previous_epoch) {
+                previous_epoch_participation.process_active_validator(
+                    val_index,
+                    state,
+                    RelativeEpoch::Previous,
+                )?;
+            }
+
+            // Note: a validator might still be "eligible" whilst returning `false` to
+            // `Validator::is_active_at`.
+            if state.is_eligible_validator(val_index)? {
```

Comment:
While looking into this I found a slight deviation from the spec in our phase0 impl of validator statuses:

https://github.com/sigp/lighthouse/blob/40b4ac374b279fd1856f1702e97b8b245f1b9bab/consensus/state_processing/src/per_epoch_processing/base/validator_statuses.rs#L202-L203

It only applies at `current_epoch = 0`, `withdrawable_epoch = 1`, in which case `!(current_epoch >= withdrawable_epoch) = true` but `previous_epoch + 1 < withdrawable_epoch = false`.

This is unreachable in practice, but I think it's worth fixing anyway by making `BeaconState::is_eligible_validator` work for all state variants, and then removing:

https://github.com/sigp/lighthouse/blob/40b4ac374b279fd1856f1702e97b8b245f1b9bab/consensus/state_processing/src/per_epoch_processing/base/rewards_and_penalties.rs#L288-L291

Reply:
Nice catch!